### DISCUSSION TITLE
How to reduce the size of libc++_shared.so

### DISCUSSION THREAD

**Discussion by FelixZhang00** [android/ndk#1661] [Q&A]
I found the libc++_shared.so in instagram.lite is very small (just 33KB)
![image](https://user-images.githubusercontent.com/5494156/151362042-7e019d69-b546-4292-ada6-e60541202f37.png)
but my simple demo's is very big(297KB)
![image](https://user-images.githubusercontent.com/5494156/151362609-8e62837b-a691-40fa-b4a3-0de94f1898c0.png)
Are there any ways to reduce the size of libc++_shared.so?

there is the flags in CmakeLists.txt:
```c++
set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -fno-exceptions -fno-rtti")

set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Oz -flto -fvisibility=hidden -fvisibility-inlines-hidden  -ffunction-sections -fdata-sections")
set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -Oz -flto -fvisibility=hidden -fvisibility-inlines-hidden -ffunction-sections -fdata-sections")
set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} -Oz -flto -Wl,--exclude-libs,ALL -Wl,--gc-sections")
set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Oz -flto -Wl,--exclude-libs,ALL -Wl,--gc-sections")
```

**Comment by DanAlbert** [comment]
Presumably AGP isn't stripping it? You're not building it so no flags you pass will do anything. Or maybe Instagram is building their own.

You only have one native library. Just use the static STL and let the linker strip the dead code. The shared STL is only needed if you have more than one library.

**Reply by enh-google** [reply]
(it won't make a 10x difference, but note also that you're comparing instagram's *compressed thumb* code for arm32 with your uncompressed arm64 code... that can make up to a 2x difference in size alone, though there's nothing you can do about it because there is no compressed form for arm64.)

**Reply by FelixZhang00** [reply]
> (it won't make a 10x difference, but note also that you're comparing instagram's _compressed thumb_ code for arm32 with your uncompressed arm64 code... that can make up to a 2x difference in size alone, though there's nothing you can do about it because there is no compressed form for arm64.)

thank you for your replay.
there is the size for arm32
![image](https://user-images.githubusercontent.com/5494156/151463367-32f998a5-8f6d-47f1-a8ef-bee4f7651ff4.png)

**Reply by FelixZhang00** [reply]
> 
thank you for your replay.

it's striped
```sh
stripped_native_libs/release/out/lib/armeabi-v7a  file libc++_shared.so 
libc++_shared.so: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=99775576f1f014c0722179fa8ecab83d1ad7e81b, stripped
```
My real project have many native libraries, so I need to use the shared STL.
I think Instagram is building their own, any ideas about which parts of the STL can be removed?

**Reply by DanAlbert** [reply]
If there was a generic answer we wouldn't ship those parts in the first place. The answer will be different for every app.
